chain/index/chain_indexer.go

```diff
@@ -0,0 +1,723 @@
+		if err != nil {
+			return xerrors.Errorf("prepare statement [%s]: %w", query, err)
+		}
+	}
+
+	return nil
+}
+
+func (ci *ChainIndexer) Close() error {
+	if ci.db == nil {
+		return nil
+	}
+	ci.cancel()
+	ci.wg.Wait()
+	return ci.db.Close()
+}
+
+func (ci *ChainIndexer) notifyUpdateSubs() {
+	ci.mu.Lock()
+	tSubs := make([]*updateSub, 0, len(ci.updateSubs))
+	for _, tSub := range ci.updateSubs {
+		tSubs = append(tSubs, tSub)
+	}
+	ci.mu.Unlock()
+
+	for _, tSub := range tSubs {
+		tSub := tSub
+		select {
+		case tSub.ch <- ChainIndexUpdated{}:
+		case <-tSub.ctx.Done():
+			// subscription was cancelled, ignore
+		case <-ci.ctx.Done():
+			return
+		}
+	}
+}
+
+func (ci *ChainIndexer) Apply(ctx context.Context, from, to *types.TipSet) error {
+	// We're moving the chain ahead from the `from` tipset to the `to` tipset
+	// Height(to) > Height(from)
+	tx, err := ci.db.BeginTx(ctx, nil)
+	if err != nil {
+		return xerrors.Errorf("beginning transaction: %w", err)
+	}
+	defer func() {
+		_ = tx.Rollback()
+	}()
+
+	// index the `to` tipset first as we only need to index the tipsets and messages for it
+	if err := ci.indexTipset(ctx, tx, to); err != nil {
+		return xerrors.Errorf("error indexing tipset: %w", err)
+	}
+
+	// index the `from` tipset in case we've not seen it before
+	if err := ci.indexTipset(ctx, tx, from); err != nil {
+		return xerrors.Errorf("error indexing tipset messages: %w", err)
+	}
+
+	// insert events for `from` tipset as it's messages have now been executed in `to` tipset
+	if err := ci.indexEvents(ctx, tx, from, to); err != nil {
+		return xerrors.Errorf("error indexing events: %w", err)
+	}
+
+	fromTsKeyCidBytes, err := toTipsetKeyCidBytes(from)
+	if err != nil {
+		return xerrors.Errorf("error getting tipset key cid: %w", err)
+	}
+
+	// events processed for `from` tipset
+	if _, err := tx.Stmt(ci.stmts.stmtMarkTipsetEventsProcessed).ExecContext(ctx, fromTsKeyCidBytes); err != nil {
+		return xerrors.Errorf("error marking tipset events processed: %w", err)
+	}
+
+	if err := tx.Commit(); err != nil {
+		return xerrors.Errorf("error committing transaction: %w", err)
+	}
+	ci.notifyUpdateSubs()
+	return nil
+}
+
+func (ci *ChainIndexer) indexEvents(ctx context.Context, tx *sql.Tx, msgTs *types.TipSet, executionTs *types.TipSet) error {
+	// check if we have an event indexed for any message in the `msgTs` tipset -> if so, there's nothig to do here
+	msgTsKeyCidBytes, err := toTipsetKeyCidBytes(msgTs)
+	if err != nil {
+		return xerrors.Errorf("error getting tipset key cid: %w", err)
+	}
+
+	// if we've already indexed events for this tipset, mark them as unreverted and return
+	res, err := tx.Stmt(ci.stmts.stmtEventsUnRevert).ExecContext(ctx, msgTsKeyCidBytes)
+	if err != nil {
+		return xerrors.Errorf("error unreverting events for tipset: %w", err)
+	}
+	rows, err := res.RowsAffected()
+	if err != nil {
+		return xerrors.Errorf("error unreverting events for tipset: %w", err)
+	}
+	if rows > 0 {
```

Comment:
ok, so we're using this as our indicator of whether we've seen these events before and won't bother checking individual events like we do now, this seems sensible I think and avoids overhead, although it has more confidence that we have a consistent state!